besu/src/main/java/org/hyperledger/besu/services/BesuPluginContextImpl.java

```diff
@@ -126,13 +128,43 @@ private void addPluginVersion(final BesuPlugin plugin) {
     pluginVersions.add(pluginVersion);
   }
 
-  public void startPlugins() {
+  public void beforeExternalServices() {
     checkState(
         state == Lifecycle.REGISTERED,
         "BesuContext should be in state %s but it was in %s",
         Lifecycle.REGISTERED,
         state);
-    state = Lifecycle.STARTING;
+    state = Lifecycle.BEFORE_EXTERNAL_SERVICES_STARTED;
+    final Iterator<BesuPlugin> pluginsIterator = plugins.iterator();
+
+    while (pluginsIterator.hasNext()) {
+      final BesuPlugin plugin = pluginsIterator.next();
+
+      try {
+        plugin.beforeExternalServices();
+        LOG.debug(
+            "beforeExternalServices called on plugin of type {}.", plugin.getClass().getName());
+      } catch (final Exception e) {
+        LOG.error(
+            "Error calling `beforeExternalServices` on plugin of type "
+                + plugin.getClass().getName()
+                + ", stop will not be called.",
```

Comment:
what does it mean "stop will not be called"

Reply:
The plugin doesn't get added to the list of running plugins, so we don't call stop when we shutdown besu. What we're saying is you need to take care of cleaning up your own resources.